Document how to create an undefined reference (Ref) and detect that using isassigned

mkitti opened the pull request:
We can create a `Ref` to a `undef` and figure out if what it points to is undefined as follows:
```
julia> struct Blah
           A
       end

julia> ref = Ref{Blah}()
Base.RefValue{Blah}(#undef)

julia> ref[]
ERROR: UndefRefError: access to undefined reference
Stacktrace:
 [1] getproperty at .\Base.jl:33 [inlined]
 [2] getindex(::Base.RefValue{Blah}) at .\refvalue.jl:32
 [3] top-level scope at REPL[16]:1

julia> isassigned(ref)
false

julia> ref[] = Blah(5)
Blah(5)

julia> isassigned(ref)
true
```

However, for primitive types the above does not work:

```
julia> ref = Ref{Int64}()
Base.RefValue{Int64}(140169022152048)

julia> ref[]
140169022152048

julia> isassigned(ref)
true
```

Because this is not intuitive, I propose to document this behavior.

Edited 7/15/2020: Update main comment since we document `isassigned` rather than `isdefined` for this purpose.

mkitti committed:
Document how to create a Ref to undef and how to detect is a Ref refers
to undef

mkitti committed:
Document Ref{T}() behavior when T is a primitve bit type.

yuyichao commented on base/refpointer.jl:
```diff
@@ -13,6 +13,11 @@ Creation of a `Ref` to a value `x` of type `T` is usually written `Ref(x)`.
 Additionally, for creating interior pointers to containers (such as Array or Ptr),
 it can be written `Ref(a, i)` for creating a reference to the `i`-th element of `a`.
 
+A `Ref` of a non-primitive type `T` referring to `undef` can be made by
```
Non bitstype.

mkitti replied:
Changed in 28f8806

yuyichao commented on base/refpointer.jl:
```diff
@@ -13,6 +13,11 @@ Creation of a `Ref` to a value `x` of type `T` is usually written `Ref(x)`.
 it can be written `Ref(a, i)` for creating a reference to the `i`-th element of `a`.
 
+A `Ref` of a non-primitive type `T` referring to `undef` can be made by
+`Ref{T}()`.  To check if a `Ref` refers to undef, use `isdefined(ref,1)`.
```
`isassigned`.

mkitti replied:
Changed in 28f8806

mkitti committed:
Change primitive to bitstype, isdefined to isassigned, doc isassigned

mkitti commented on base/essentials.jl:
```diff
@@ -623,6 +623,10 @@ getindex(v::SimpleVector, I::AbstractArray) = Core.svec(Any[ v[i] for i in I ]..
 Test whether the given array has a value associated with index `i`. Return `false`
 if the index is out of bounds, or has an undefined reference.
 
+    isassigned(ref) -> Bool
+
+Test whether the given [`Ref`](@ref) has been associated with a value. This is always true for a [`Ref`](@ref) for a bitstype. Return `false` if the reference is undefined.
+
```
Added `isassigned(ref)` documentation with cross reference to `Ref`

yuyichao replied:
I think this (and the example) should add to the docstring of `isassigned(::RefValue)` instead. You can use `methods(isassigned)` (or equivalent) to find where it is.

mkitti replied:
Done in 2849f95

mkitti commented on base/essentials.jl:
```diff
@@ -641,6 +645,18 @@ julia> v = similar(rand(3), Foo)
 
 julia> isassigned(v, 1)
 false
+
+julia> ref = Ref{Function}()
+Base.RefValue{Function}(#undef)
+
+julia> ref[] = max
+max (generic function with 12 methods)
+
+julia> isassigned(ref)
+true
+
+julia> isassigned(Ref{Int}())
+true
```
Added examples of `isassigned` for `Ref`

yuyichao commented on base/refpointer.jl:
```diff
@@ -13,6 +13,11 @@ Creation of a `Ref` to a value `x` of type `T` is usually written `Ref(x)`.
+
 When passed as a `ccall` argument (either as a `Ptr` or `Ref` type), a `Ref` object will be
 converted to a native pointer to the data it references.
 
```
This should probably be merged with the paragraph about `NULL` right below. The mention of `NULL` below is also wrong for exactly the case mentioned above...

mkitti replied:
I removed the no invalid reference language in 2849f95

mkitti committed:
Merge branch 'master' into isdefined_ref

mkitti committed:
Document isassigned(::RefValue). Remove no invalid Ref language.

mkitti commented:
I have made the changes suggested. With the phrase about no invalid `Refs` existing removed, the remaining sentence appears to be more about the use of `C_NULL` with `ccall`

mkitti commented:
This just came up in Slack helpdesk again: https://julialang.slack.com/archives/C6A044SQH/p1594629154297600

@yuyichao What is needed to be done on this PR?

yuyichao commented:
The duplicate should be removed from the docstring for `isassigned` for arrays.

mkitti committed:
Merge branch 'master' into isdefined_ref

mkitti committed:
Remove duplicate isassigned docs for arrays in essentials.jl

yuyichao commented on base/refpointer.jl:
```diff
@@ -13,11 +13,20 @@ Creation of a `Ref` to a value `x` of type `T` is usually written `Ref(x)`.
 
-There is no invalid (NULL) `Ref` in Julia, but a `C_NULL` instance of `Ptr` can be passed to
-a `ccall` Ref argument.
+To check if a `Ref` refers to [`undef`](@ref), use [`isassigned(ref::RefValue)`](@ref).
```
I highly suspect that this will link to `undef`, which is wrong. It's a missing reference, it's not refering to any object. A reference to `undef` is `Ref(undef)`.

mkitti replied:
I just say "undefined reference" now in 94edfaa

mkitti committed:
Refer to undefined reference rather than undef for isassigned

yuyichao reviewed (approved):
There are some super long lunges but lgtm otherwise.

mkitti committed:
isassigned reference docs, fix line returns in doc

mkitti commented:
I fixed "super long lunges" in f9fbfa2

mkitti committed:
Fix line returns for isassigned(RefValue) doc

rfourquet commented on base/refpointer.jl:
```diff
@@ -13,11 +13,20 @@ Creation of a `Ref` to a value `x` of type `T` is usually written `Ref(x)`.
-converted to a native pointer to the data it references.
+`Ref{T}()` creates a reference to a value of type `T` without initialization.
+For a bitstype `T`, the value will be whatever currently resides in the memory
+allocated. For a non-bitstype `T`, the value will be `undef` and will result in
```
This seems confusing, `undef` represents a well-defined singleton value in Julia. I would just write "undefined". Also, it's not the value which will result in an error, but the act of dereferencing.

mkitti replied:
I corrected this in a49c8cc

rfourquet replied:
Sorry for the delay. If I'm not mistaken, my second point was not addressed. That is, the reference will not "result in an error"; trying to dereference it will result in an error.

mkitti replied:
I revised the text in 2b9a7a7 to clarify that attempted dereferencing will cause an error. Thanks @rfourquet

rfourquet commented on base/refpointer.jl:
```diff
@@ -13,11 +13,20 @@ Creation of a `Ref` to a value `x` of type `T` is usually written `Ref(x)`.
-There is no invalid (NULL) `Ref` in Julia, but a `C_NULL` instance of `Ptr` can be passed to
-a `ccall` Ref argument.
+To check if a `Ref` is an undefined reference, use [`isassigned(ref::RefValue)`](@ref).
+For example, `isassigned(Ref{T}())` is `false` if T is not a bitstype struct.
```
```suggestion
For example, `isassigned(Ref{T}())` is `false` if `T` is not a bitstype.
```

rfourquet commented on base/refpointer.jl:
```diff
@@ -13,11 +13,20 @@ Creation of a `Ref` to a value `x` of type `T` is usually written `Ref(x)`.
-a `ccall` Ref argument.
+To check if a `Ref` is an undefined reference, use [`isassigned(ref::RefValue)`](@ref).
+For example, `isassigned(Ref{T}())` is `false` if T is not a bitstype struct.
+If T is a bitstype, `isassigned(Ref{T}())` will always be true.
```
```suggestion
If `T` is a bitstype, `isassigned(Ref{T}())` will always be true.
```

rfourquet commented on base/refpointer.jl:
```diff
@@ -13,11 +13,20 @@ Creation of a `Ref` to a value `x` of type `T` is usually written `Ref(x)`.
+When passed as a `ccall` argument (either as a `Ptr` or `Ref` type), a `Ref`
+object will be converted to a native pointer to the data it references.
+
+A `C_NULL` instance of `Ptr` can be passed to a `ccall` Ref argument to initialize it.
```
```suggestion
A `C_NULL` instance of `Ptr` can be passed to a `ccall` `Ref` argument to initialize it.
```

rfourquet commented on base/refvalue.jl:
```diff
@@ -8,6 +8,31 @@ mutable struct RefValue{T} <: Ref{T}
+    isassigned(ref::RefValue) -> Bool
+
+Test whether the given [`Ref`](@ref) has been associated with a value.
+This is always true for a [`Ref`](@ref) for a bitstype.
```
(to avoid "for" repetition... feel free to ignore)
```suggestion
This is always true for a [`Ref`](@ref) of a bitstype object.
```

rfourquet commented on base/refvalue.jl:
```diff
@@ -8,6 +8,31 @@ mutable struct RefValue{T} <: Ref{T}
+"""
+    isassigned(ref::RefValue) -> Bool
+
+Test whether the given [`Ref`](@ref) has been associated with a value.
```
```suggestion
Test whether the given [`Ref`](@ref) is associated with a value.
```

mkitti committed:
Add backticks to quote code as suggested by @rfourquet

Co-authored-by: Rafael Fourquet <fourquet.rafael@gmail.com>

mkitti committed:
Change `undef` to undefined, fix examples

Change `undef` to undefined reference

Add comments to examples
Remove C_NULL example due to confusion
List normal `Ref` usage first

mkitti commented on base/refpointer.jl:
```diff
@@ -28,6 +37,18 @@ julia> isa.(Ref([1,2,3]), [Array, Dict, Int])
+julia> isassigned(Ref{Int64}())
+true
+
+julia> Ref(C_NULL)
```
This was supposed to show "normal" usage of `Ref` with a value, but it ends up being more confusing. I created an example just using `Ref(5)` and put it first.

mkitti committed:
Create a markdown header for examples

mkitti commented:
Thank you for all of the reviews. Are there any other pending issues to resolve, @KristofferC or @rfourquet?

mkitti committed:
Clarify that error when attempting to deference

Clarify that the act of dereferencing causes the error to be thrown. @rfourquet

rfourquet reviewed (approved):
LGTM, modulo maybe this test which might fail sometimes.

mkitti committed:
Remove undefined Ref bitstype comparison

Comparison of an undefined reference value bitstype to zero may be true. Insert a semicolon just to demonstrate usage but not the result.

mkitti commented on base/refpointer.jl:
```diff
@@ -13,21 +13,53 @@ Creation of a `Ref` to a value `x` of type `T` is usually written `Ref(x)`.
+       end
+UndefRefError()
+
+julia> Ref{Int64}()[] == 0 # A reference to a bitstype refers to an undetermined value if not given
```
Instead of doing the comparison, I just added a semicolon in 08713b2. We'll just demonstrate the lack of an error and not the undetermined output.

mkitti commented on doc/src/base/c.md:
```diff
@@ -23,6 +23,7 @@ Base.systemerror
 Base.windowserror
 Core.Ptr
 Core.Ref
+Base.isassigned(::Base.RefValue)
```
I added this to `c.md` next to where `Ref` is introduced.. Is this the right place to insert this in the docs? Do we need additional docs on `Base.RefValue` ?

mkitti commented:
Hello everyone. I'm relocating over the next week or so and will be unavailable.

Please let me know if there is anything more to do on this doc PR. Thanks.

mkitti commented:
I'm now back online.

mkitti commented:
@KristofferC, is there anything left to do on this doc PR?

mkitti commented:
Hello. Just checking in on this documentation-only PR. @KristofferC was there anything you wanted to revise on this doc PR?
